### ISSUE TITLE
Optional arguments

### ISSUE THREAD

**Issue by JorelAli** [JorelAli/CommandAPI#162]
**Please describe your suggestion**
Optional arguments

**Please supply examples of the desired inputs and outputs**
_Normal usage:_
User enters this:
```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withOptionalArgument(new IntegerArgument("optional"), 10)
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1]; // This is 10 if not existant
  })
  .register();
```

Basically gets converted into this:
```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withArgument(new IntegerArgument("optional"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1];
  })
  .register();
```
and
```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1]; // This is 10, args.length = 2
  })
  .register();
```

_Using annotations:_
Either something like this:
```java
@Subcommand("mycommand")
@Arg(name = "int", type = IntegerArgument.class)
@OptionalArg(name = "int", type = IntegerArgument.class, optional = "10")
public void onCommand(int arg1, int arg2) {
  //...
}
```

or something like this:
```java
@Subcommand("mycommand")
@Arg(name = "int", type = IntegerArgument.class)
@Arg(name = "optional", type = IntegerArgument.class)
public void onCommand(int arg1, @Default("10") int arg2) {
  //...
}
```

Cons of this via annotations is that the value must be a string, primitive, enum, annotation or class, which is really restrictive.

**Comment by SNWCreations** [comment]
Emmm, so is it really implemented?
I like this feature, but I can't find it in v7.0 :(

**Comment by SNWCreations** [comment]
Emmm, so is it really implemented?
I like this feature, but I can't find it in v7.0 :(

**Comment by JorelAli** [comment]
@SNWCreations No, this feature is not implemented (yet?)

**Comment by SNWCreations** [comment]
> 

Emmm, that's too bad :(
If so, will it be available in v7.x? (I'm still working on 1.16.5, I need it!)

**Comment by JorelAli** [comment]
@SNWCreations
This feature, despite how useful it is, is not relatively high on priority at the moment. This is because this feature can be implemented using other methods, for example, something like this:

```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withOptionalArgument(new IntegerArgument("optional"), 10)
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1]; // This is 10 if not existant
  })
  .register();
```

Can already be implemented as this:

```java
public void execute(CommandSender sender, int arg1, int arg2) {
  // Command logic here
}
```

And then you implement your CommandAPICommands as follows:
```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    execute(sender, arg1, 10);
  })
  .register();

new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withArgument(new IntegerArgument("optional"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1];
    execute(sender, arg1, arg2);
  })
  .register();
```

-----

This feature is also surprisingly difficult to implement and ends up causing more ambiguity and issues than explicitly stating everything manually.

**Comment by SNWCreations** [comment]
> @SNWCreations This feature, despite how useful it is, is not relatively high on priority at the moment. This is because this feature can be implemented using other methods, for example, something like this:
> 
> ```java
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .withOptionalArgument(new IntegerArgument("optional"), 10)
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     int arg2 = (int) args[1]; // This is 10 if not existant
>   })
>   .register();
> ```
> 
> Can already be implemented as this:
> 
> ```java
> public void execute(CommandSender sender, int arg1, int arg2) {
>   // Command logic here
> }
> ```
> 
> And then you implement your CommandAPICommands as follows:
> 
> ```java
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     execute(sender, arg1, 10);
>   })
>   .register();
> 
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .withArgument(new IntegerArgument("optional"))
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     int arg2 = (int) args[1];
>     execute(sender, arg1, arg2);
>   })
>   .register();
> ```
> 
> Optional arguments is also surprisingly difficult to implement and ends up causing more ambiguity and issues than explicitly stating everything manually.

Can I use it in annotation framework?

**Comment by JorelAli** [comment]
> Can I use it in annotation framework?

No.

**Comment by SNWCreations** [comment]
This feature, despite how useful it is, is not relatively high on priority at the moment. This is because this feature can be implemented using other methods, for example, something like this:

new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withOptionalArgument(new IntegerArgument("optional"), 10)
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1]; // This is 10 if not existant
  })
  .register();
Can already be implemented as this:

public void execute(CommandSender sender, int arg1, int arg2) {
  // Command logic here
}

My reply: I can't understand this. :(

**Comment by JorelAli** [comment]
@SNWCreations 

This code:

```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withOptionalArgument(new IntegerArgument("optional"), 10)
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1]; // This is 10 if not existant
  })
  .register();
```

Is the same as this code:

```java
new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    execute(sender, arg1, 10);
  })
  .register();

new CommandAPICommand("blah")
  .withArgument(new IntegerArgument("int"))
  .withArgument(new IntegerArgument("optional"))
  .executes((sender, args) -> {
    int arg1 = (int) args[0];
    int arg2 = (int) args[1];
    execute(sender, arg1, arg2);
  })
  .register();
  
public void execute(CommandSender sender, int arg1, int arg2) {
  // ...
}

```

**Comment by SNWCreations** [comment]
> @SNWCreations
> 
> This code:
> 
> ```java
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .withOptionalArgument(new IntegerArgument("optional"), 10)
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     int arg2 = (int) args[1]; // This is 10 if not existant
>   })
>   .register();
> ```
> 
> Is the same as this code:
> 
> ```java
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     execute(sender, arg1, 10);
>   })
>   .register();
> 
> new CommandAPICommand("blah")
>   .withArgument(new IntegerArgument("int"))
>   .withArgument(new IntegerArgument("optional"))
>   .executes((sender, args) -> {
>     int arg1 = (int) args[0];
>     int arg2 = (int) args[1];
>     execute(sender, arg1, arg2);
>   })
>   .register();
>   
> public void execute(CommandSender sender, int arg1, int arg2) {
>   // ...
> }
> ```

Thanks! I got it.

**Comment by JorelAli** [comment]
It would be very good to get this implemented (or at least have made significant progress on this) for a 8.5.0 release due late July or early August.

**Comment by SNWCreations** [comment]
> After just over 2 years, this has now been implemented and merged into the upcoming 9.0.0 release.

I'm so happy to see this. :D

**Comment by JorelAli** [comment]
Implemented in 9.0.0.
